lib/active_record/associations/has_many_split_through_association.rb

```diff
@@ -4,28 +4,31 @@ module ActiveRecord
   module Associations
     class SplitAssociationScope < AssociationScope
       def scope(association)
+        # source of the through reflection
         reflection = association.reflection
+        #remove all previously set scope of passed in association
         scope = association.klass.unscoped
 
         chain = get_chain(reflection, association, scope.alias_tracker)
 
-        join_ids = [association.owner.id]
-        records = nil
-
         reverse_chain = chain.reverse
-        last_reflection = reverse_chain.last
-        reverse_chain.each do |refl|
-          records = refl.klass.unscoped.where(refl.join_keys.key => join_ids)
+        first_reflection = reverse_chain.shift
+        first_join_ids = [association.owner.id]
+        initial_values = [first_reflection, first_join_ids]
+
+        last_reflection, join_ids = reverse_chain.inject(initial_values) do |(reflection, join_ids), next_reflection|
+          key = reflection.join_keys.key
+          records = reflection.klass.where(key => join_ids)
+
           # Preventing the reflection from being loaded on the
           # last reflection in the chain, that way anything the user
           # wants to apply to the reflection will still work.
-          if refl != last_reflection
-            records = records.select(:id)
-            join_ids = records.map(&:id)
-          end
+          foreign_key = next_reflection.join_keys.foreign_key
+          [next_reflection, records.pluck(foreign_key)]
         end
 
-        records
+        key = last_reflection.join_keys.key
+        last_reflection.klass.where(key => join_ids)
```

Comment:
```suggestion
        last_reflection.klass.where(key => last_join_ids)
```

...and then using the new `last_` name here.